### ISSUE TITLE
Deploy error (minifier breaks on some code)

### ISSUE THREAD

**Issue by steph643** [meteor/windows-preview#104]
I use v0.3.0 and I get the following error when running 'meteor deploy':

```
C:\Users\Steph\Documents\Travail\Meteor\myapp>meteor deploy myapp.meteor.com
Deploying to myapp.meteor.com.

C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\mt-os.wi
ndows.x86_32\dev_bundle\lib\node_modules\fibers\future.js:245
                                                throw(ex);
                                                      ^
Error: Unexpected token: punc ()) (line: 32, col: 1, pos: 1785)
    at Object.Future.wait (C:\Users\Steph\AppData\Local\.meteor\packages\meteo
r-tool\1.1.1-win.7\mt-os.windows.x86_32\dev_bundle\lib\node_modules\fibers\futur
e.js:395:16)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:556:31
    at Array.forEach (native)
    at Function._.each._.forEach (C:\Users\Steph\AppData\Local\.meteor\package
s\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\dev_bundle\lib\node_modules\under
score\underscore.js:79:11)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:554:9
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:352:18
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:345:34
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:343:23
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at enterJob (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1
.1-win.7\mt-os.windows.x86_32\tools\buildmessage.js:317:26)
    at Object.forkJoin (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-t
ool\1.1.1-win.7\mt-os.windows.x86_32\tools\buildmessage.js:531:3)
    at [object Object]._.extend.minifyJs (C:\Users\Steph\AppData\Local\.meteor
\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\bundler.js:818:28)
    at [object Object]._.extend.make (C:\Users\Steph\AppData\Local\.meteor\pac
kages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\bundler.js:509:12)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\bundler.js:2054:14
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\bundler.js:2099:20
    at Array.forEach (native)
    at Function._.each._.forEach (C:\Users\Steph\AppData\Local\.meteor\package
s\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\dev_bundle\lib\node_modules\under
score\underscore.js:79:11)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\bundler.js:2098:7
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:264:13
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:257:29
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:255:18
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:246:23
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at Object.capture (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-to
ol\1.1.1-win.7\mt-os.windows.x86_32\tools\buildmessage.js:245:19)
    at Object.exports.bundle (C:\Users\Steph\AppData\Local\.meteor\packages\me
teor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\bundler.js:2040:31)
    at Object.bundleAndDeploy (C:\Users\Steph\AppData\Local\.meteor\packages\m
eteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\deploy.js:405:32)
    at Command.main.registerCommand.name [as func] (C:\Users\Steph\AppData\Loc
al\.meteor\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\commands.
js:1146:29)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\main.js:1345:23
    - - - - -
    at Object.Future.wait (C:\Users\Steph\AppData\Local\.meteor\packages\meteo
r-tool\1.1.1-win.7\mt-os.windows.x86_32\dev_bundle\lib\node_modules\fibers\futur
e.js:395:16)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:556:31
    at Array.forEach (native)
    at Function._.each._.forEach (C:\Users\Steph\AppData\Local\.meteor\package
s\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\dev_bundle\lib\node_modules\under
score\underscore.js:79:11)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:554:9
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:352:18
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:345:34
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:343:23
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at enterJob (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1
.1-win.7\mt-os.windows.x86_32\tools\buildmessage.js:317:26)
    at Object.forkJoin (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-t
ool\1.1.1-win.7\mt-os.windows.x86_32\tools\buildmessage.js:531:3)
    at _minify (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.
1-win.7\mt-os.windows.x86_32\tools\bundler.js:931:18)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\bundler.js:820:16
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:537:23
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:352:18
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:345:34
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:343:23
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at enterJob (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1
.1-win.7\mt-os.windows.x86_32\tools\buildmessage.js:317:26)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:536:24
    at runWithEnvironment (C:\Users\Steph\AppData\Local\.meteor\packages\meteo
r-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:139:21)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\fiber-helpers.js:146:14
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:549:11
    - - - - -

Error
    at new JS_Parse_Error (C:\Users\Steph\AppData\Local\.meteor\packages\meteo
r-tool\1.1.1-win.7\mt-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_
modules\uglify-js\lib\parse.js:196:18)
    at js_error (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1
.1-win.7\mt-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_modules\ug
lify-js\lib\parse.js:204:11)
    at croak (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-
win.7\mt-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_modules\uglif
y-js\lib\parse.js:674:9)
    at token_error (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\
1.1.1-win.7\mt-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_modules
\uglify-js\lib\parse.js:682:9)
    at unexpected (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1
.1.1-win.7\mt-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_modules\
uglify-js\lib\parse.js:688:9)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_modules\uglify-js\li
b\parse.js:778:17
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_modules\uglify-js\li
b\parse.js:721:24
    at block_ (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1
-win.7\mt-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_modules\ugli
fy-js\lib\parse.js:1001:20)
    at ctor.body (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.
1.1-win.7\mt-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_modules\u
glify-js\lib\parse.js:974:25)
    at function_ (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.
1.1-win.7\mt-os.windows.x86_32\isopackets\minifiers\npm\minifiers\node_modules\u
glify-js\lib\parse.js:979:15)
    at Object.Future.get (C:\Users\Steph\AppData\Local\.meteor\packages\meteor
-tool\1.1.1-win.7\mt-os.windows.x86_32\dev_bundle\lib\node_modules\fibers\future
.js:196:86)
    at Object.Future.wait (C:\Users\Steph\AppData\Local\.meteor\packages\meteo
r-tool\1.1.1-win.7\mt-os.windows.x86_32\dev_bundle\lib\node_modules\fibers\futur
e.js:395:16)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:556:31
    at Array.forEach (native)
    at Function._.each._.forEach (C:\Users\Steph\AppData\Local\.meteor\package
s\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\dev_bundle\lib\node_modules\under
score\underscore.js:79:11)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:554:9
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:352:18
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:345:34
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:343:23
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at enterJob (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1
.1-win.7\mt-os.windows.x86_32\tools\buildmessage.js:317:26)
    at Object.forkJoin (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-t
ool\1.1.1-win.7\mt-os.windows.x86_32\tools\buildmessage.js:531:3)
    at _minify (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.
1-win.7\mt-os.windows.x86_32\tools\bundler.js:931:18)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\bundler.js:820:16
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:537:23
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:352:18
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:345:34
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:343:23
    at [object Object]._.extend.withValue (C:\Users\Steph\AppData\Local\.meteo
r\packages\meteor-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:1
15:14)
    at enterJob (C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1
.1-win.7\mt-os.windows.x86_32\tools\buildmessage.js:317:26)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:536:24
    at runWithEnvironment (C:\Users\Steph\AppData\Local\.meteor\packages\meteo
r-tool\1.1.1-win.7\mt-os.windows.x86_32\tools\fiber-helpers.js:139:21)
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\fiber-helpers.js:146:14
    at C:\Users\Steph\AppData\Local\.meteor\packages\meteor-tool\1.1.1-win.7\m
t-os.windows.x86_32\tools\buildmessage.js:549:11
C:\Users\Steph\Documents\Travail\Meteor\myapp>
```

**Comment by steph643** [comment]
Error occurs during "minifying"

**Comment by steph643** [comment]
Guess I got a ")" somewhere that bothers the minifyer, but how to find it?

**Comment by steph643** [comment]
Same error with `meteor --production`

**Comment by steph643** [comment]
This makes the minifyer happy:

```javascript
function fieldIsRequired(fieldSchema)
    {
/*
    // If schema contains 'optional', then field is not mandatory
//    if (fieldSchema.optional)
//        return false;

    // If schema type is String and there is no 'min' in the schema (or min===0), then field is not mandatory
    if (fieldSchema.type === String && !fieldSchema.min)
        return false;

 //   return true;
*/
    }
```
This makes the minifyer trigger the above error:
```javascript
function fieldIsRequired(fieldSchema)
    {
    // If schema contains 'optional', then field is not mandatory
//    if (fieldSchema.optional)
//        return false;
/*
    // If schema type is String and there is no 'min' in the schema (or min===0), then field is not mandatory
    if (fieldSchema.type === String && !fieldSchema.min)
        return false;
*/
 //   return true;
    }
```
I am scared.

**Comment by stubailo** [comment]
@steph643 - I have just created a new Meteor project and pasted in both code snippets, and the minifier ran fine. It would be super helpful if you could provide an example of a complete project that causes the minifier to break.

**Comment by avital** [comment]
(And is this special to Windows?)

**Comment by steph643** [comment]
@stubailo, thanks a lot. I will investigate more. There should be some kind of file format issue, I guess. I got the error in another place too. I will try to see what is common to the two places.
@avital, I will try on MacOS in the coming hours.

**Comment by steph643** [comment]
Got it. Here is the minimal text that breaks the minifyer, **given that you use 0D as the line break characters** instead of 0A: 
```javascript
{
//
}
```
So this is ok:
```
7B 0A 2F 2F 0A 7D                 {.//.}
```
And this is not:
```
7B 0D 2F 2F 0D 7D                 {.//.}
```

I don't know how I ended up having 0D characters in my files. Did I do something wrong or is it a Meteor bug?

**Comment by Slava** [comment]
@steph643 I guess it is windows line terminator but it is missing the paired 0A for some reason? http://homepage.smc.edu/morgan_david/CS41/lineterminators.htm

**Comment by stubailo** [comment]
I am guessing `0D` is `\r`?

**Comment by Slava** [comment]
@steph643 out of curiosity, what text editor did you use to construct this text?

**Comment by stubailo** [comment]
I think we can add an option to `read` in build plugins to convert line endings? Then add a corresponding option to `files.readFile`

**Comment by steph643** [comment]
@Slava, my first files have been created (and converted to "UTF8 without BOM") with Notepad++ under Windows. Then they have been edited (and some more have been created) with WebStorm under MacOS and Windows in turns (as my source code is shared on both OSs through Dropbox).

**Comment by steph643** [comment]
A quick look at my files shows they are mostly using single 0D. But I can also see some 0D 0A.

What is amazing is that I also have a single 0A, which I have just created for my post above but cannot create again!

**Comment by Slava** [comment]
I think overall we should just normalize files to replace `\r\n` and `\r` to `\n`, and provide a method to build plugins to override it.

**Comment by steph643** [comment]
@avital: no issue under MacOS with neither of the three line break formats.

**Comment by steph643** [comment]
This issue doesn't only trigger the exception described at the top of this thread. It also breaks the minifier in more silent ways. As far as I can tell, the resulting broken code mainly has missing global variables.

I solved this by doing a bulk EOL conversion under WebStorm: select the root folder in the tree view, then File > Line Separators (I have saved my broken project to test the fix).

**Comment by Slava** [comment]
Should be fixed on a branch! Thank you for a clear reproduction. I successfully reproduced and verified that my fix addresses the problem.

**Comment by stubailo** [comment]
Will close when the fix is released

**Comment by Slava** [comment]
Fixed in 1.1 RC0
